README.md
# Shoten

[![Python package](https://img.shields.io/pypi/v/shoten.svg)](https://pypi.python.org/pypi/shoten)
[![Python versions](https://img.shields.io/pypi/pyversions/shoten.svg)](https://pypi.python.org/pypi/shoten)
[![Code Coverage](https://img.shields.io/codecov/c/github/adbar/shoten.svg)](https://codecov.io/gh/adbar/shoten)
[![image](https://zenodo.org/badge/DOI/10.5281/zenodo.6365800.svg)](https://doi.org/10.5281/zenodo.6365800)

Helper functions to find word trends (i.e. extract tokens, lemmatize and filter).


## Installation

With `pip` (`pip3` where applicable):

- Package repository: `pip install shoten`
- Latest development version:
  `pip install -U git+https://github.com/adbar/shoten.git`


## Usage

### Input

Two possibilities for input data:

- XML-TEI files as generated by [trafilatura](https://trafilatura.readthedocs.io/):
   1. `from shoten import gen_wordlist`
   2. `vocab = gen_wordlist(mydir, ['de', 'en'])`

- TSV-file contaning a word list: word form + `TAB` + date (`YYYY-MM-DD` format) + possible 3rd column (source)
   1. `from shoten import load_wordlist`
   2. `vocab = load_wordlist(myfile, ['de', 'en'])`

Language codes: optional list of languages to be considered for
lemmatization, ordered by relevance. [ISO 639-1
codes](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes), see the
[list of supported languages](https://github.com/adbar/simplemma).

Optional argument `maxdiff`: maximum number of days to consider
(default: 1000, i.e. going back up to 1000 days from today).

### Filters

`from shoten.filters import *`

Given a vocabulary `vocab` as explained above:

- `hapax_filter(vocab, freqcount=2)`: (default frequency: \<= 2)
- `different_days_filter(vocab, threshold=2)`: word seen on at least n
  days
- `regex_filter(vocab, string)`: custom regular expression
- `compound_filter(vocab, language, threshold)`: examine parts of
  hyphenated compounds based on target language and corpus frequency
- `shortness_filter(vocab, threshold=20)`: length threshold in percent
  of word lengths
- `frequency_filter(vocab, max_perc=50, min_perc=.001)`: maximum and
  minimum frequencies in percent
- `oldest_filter(vocab, threshold=50)`: discard the oldest words
  (threshold in percent)
- `freshness_filter(vocab, percentage=10)`: keep the X% freshest words
- `ngram_filter(vocab, threshold=90, verbose=False)`: retains X% words
  based on character n-gram frequencies; runs out of memory if the
  vocabulary is too large (8 GB RAM recommended)
- `sources_freqfilter(vocab, threshold=2)`: remove words which are
  only present in less than x sources
- `sources_filter(vocab, myset)`: only keep the words for which the
  source contains a string listed in the input set
- `wordlist_filter(vocab, mylist, keep_words=False)`: keep or discard
  words present in the input list

Reduce vocabulary size with a filter:

`vocab = oldest_filter(vocab)`

Filter chains:

`vocab = oldest_filter(shortness_filter(vocab))`

### Output

``` python
# print one-by-one
for word in sorted(vocab):
    print(word)
# transfer to a list
results = [w for w in vocab]
```

### CLI

`shoten --help`

## Additional information

*Shoten* = focal point in Japanese
([焦点](https://en.wiktionary.org/wiki/%E7%84%A6%E7%82%B9#Japanese)).

Project webpage: [Webmonitor](https://www.dwds.de/d/korpora/webmonitor).


HISTORY.rst
=======
History
=======



0.1.0
-----

* First release on PyPI.


tests/__init__.py
"""Unit test package for shoten."""


tests/test_shoten.py
#!/usr/bin/env python

"""Tests for `shoten` package."""

import re
import sys
import tempfile

from array import array
from collections import Counter
from copy import deepcopy
from pathlib import Path
from unittest.mock import patch

import pytest

from shoten import *
from shoten.cli import parse_args, process_args
from shoten.filters import *

from simplemma import load_data


DE_LEMMADATA = load_data('de')


# Turns a dictionary into a class
class ToEntry(Entry):
    "Convert a dictionary to the entry class."
    def __init__(self, my_dict):
        for key in my_dict:
            setattr(self, key, my_dict[key])
        setattr(self, 'series_rel', array('f'))


def test_basics():
    """Test basic functions."""
    # dates
    assert calc_timediff('2020 A') is None
    assert calc_timediff('2020-01-01') > 1
    assert calc_timediff('2030-01-01') < 1
    # load words
    myvocab = load_wordlist(str(Path(__file__).parent / 'inputfile.txt'))
    assert len(myvocab) == 3
    assert myvocab['Other'].sources == Counter({'Source1': 1})
    # pickling and unpickling
    os_handle, filepath = tempfile.mkstemp(suffix='.pickle', text=True)
    pickle_wordinfo(myvocab, filepath)
    myvocab2 = unpickle_wordinfo(filepath)
    assert len(myvocab2) == len(myvocab) and myvocab2['Tests'].time_series.all() == myvocab['Tests'].time_series.all()
    # generate from XML file
    myvocab = gen_wordlist(str(Path(__file__).parent / 'testdir' / 'test2'), langcodes=('de'))
    assert len(myvocab) == 1 and 'Telegram' in myvocab
    # single- vs. multi-threaded
    assert gen_wordlist(str(Path(__file__).parent / 'testdir'), langcodes=('de'), threads=1).keys() == gen_wordlist(str(Path(__file__).parent / 'testdir'), langcodes=('de'), threads=3).keys()
    # generate list
    myvocab = gen_wordlist(str(Path(__file__).parent / 'testdir'), langcodes=('de', 'en'))
    assert 'Messengerdienst' in myvocab
    # without language codes and with short time frame
    myvocab = gen_wordlist(str(Path(__file__).parent / 'testdir'), langcodes=[])
    assert 'Messengerdienst' in myvocab
    # without language codes and with short time frame
    myvocab = gen_wordlist(str(Path(__file__).parent / 'testdir'), maxdiff=1)
    assert 'Messengerdienst' not in myvocab
    # with author filter
    myregex = re.compile(r'\b(|afp|apa|dpa)\b', re.I)
    myvocab = gen_wordlist(str(Path(__file__).parent / 'testdir'), authorregex=myregex)
    assert 'Messengerdienst' not in myvocab
    # test frequency calculations
    assert gen_freqlist(str(Path(__file__).parent / 'testdir')) == {}
    assert gen_freqlist(str(Path(__file__).parent / 'testdir'), langcodes=['en']) == {}
    mydict = {
        'Test': ToEntry({'total': 20, 'mean': 10, 'stddev': 0, 'series_rel': [10, 10]})
    }
    # write to temp file
    os_handle, temp_outputfile = tempfile.mkstemp(suffix='.tsv', text=True)
    result = store_freqlist(mydict, temp_outputfile)
    assert result is None
    mydict['Test2'] = ToEntry({'total': 10, 'mean': 5, 'stddev': 4.082, 'series_rel': [10, 5, 0]})
    result = store_freqlist(mydict, temp_outputfile)
    assert result is None


def test_internals():
    """Test internal functions."""
    # filter known lemmata
    assert filter_lemmaform('Berge', DE_LEMMADATA, lemmafilter=True) is None
    assert filter_lemmaform('Berge', DE_LEMMADATA, lemmafilter=False) == 'Berg'

    # store in vocabulary
    myvocab = {}
    myvocab = putinvocab(myvocab, 'Berge', 5, source='Source_0', inheadings=True)
    assert 'Berge' in myvocab and myvocab['Berge'].time_series == array('H', [5]) and myvocab['Berge'].headings is True

    # de-hyphening
    myvocab = {
        'de-hyphening': ToEntry({'time_series': array('H', [1, 2, 3]), 'sources': Counter(['source1']), 'headings': True}),
        'dehyphening': ToEntry({'time_series': array('H', [3, 4]), 'sources': Counter(['source2']), 'headings': False})
    }
    newvocab = prune_vocab(deepcopy(myvocab), 'de-hyphening', 'dehyphening')
    assert len(newvocab['dehyphening'].time_series) == 5
    assert newvocab['dehyphening'].headings is True
    newvocab = dehyphen_vocab(deepcopy(myvocab))
    assert newvocab != myvocab
    assert len(newvocab['dehyphening'].time_series) == 5
    assert newvocab['dehyphening'].headings is True

    # refine vocab
    myvocab['Berge'] = ToEntry({'time_series': array('H', [5]), 'sources': Counter(['source3']), 'headings': False})
    newvocab = refine_vocab(deepcopy(myvocab), DE_LEMMADATA, lemmafilter=False, dehyphenation=False)
    assert 'Berg' in newvocab and 'de-hyphening' in newvocab
    newvocab = refine_vocab(deepcopy(myvocab), DE_LEMMADATA, lemmafilter=True, dehyphenation=True)
    assert 'Berge' not in newvocab and 'Berg' not in newvocab and 'de-hyphening' not in newvocab

    # filter levels
    newvocab = convert_to_numpy(newvocab)
    apply_filters(deepcopy(newvocab))
    apply_filters(newvocab, setting='loose')
    with pytest.raises(ZeroDivisionError):
        apply_filters(newvocab, setting='strict')

    # frequencies
    oldestday, newestday = 21, 1
    myvocab = {
        'Berg': ToEntry({'time_series': array('H', [newestday, 10, 10, oldestday])}),
        'Meer': ToEntry({'time_series': array('H', [newestday, 8, 9, 11, oldestday])}),
        'Tal': ToEntry({'time_series': array('H', [newestday, oldestday])}),
        'Zebra': ToEntry({'time_series': array('H', [10])})
    }
    bins = calculate_bins(myvocab, interval=7)
    assert bins == [14, 7]
    myvocab = refine_frequencies(myvocab, bins)
    assert 'Berg' in myvocab and 'Tal' not in myvocab and 'Zebra' not in myvocab

    # freq calculations
    myvocab, timeseries = compute_frequencies(myvocab, bins)
    assert myvocab['Berg'].total == 400000.0 and myvocab['Berg'].series_abs == array('H', [0, 2])
    assert myvocab['Meer'].total == 600000.0 and myvocab['Meer'].series_abs == array('H', [0, 3])
    assert timeseries == [0, 5]
    myvocab = combine_frequencies(myvocab, bins, timeseries)
    assert myvocab['Berg'].series_rel == array('f', [0, 400000.0])
    assert myvocab['Meer'].series_rel == array('f', [0, 600000.0])


def test_cli():
    """Basic tests for command-line interface."""
    mydir = str(Path(__file__).parent / 'testdir')
    testargs = ['', '--read-dir', mydir, '-l', 'de', '--filter-level', 'loose']
    with patch.object(sys, 'argv', testargs):
        args = parse_args(testargs)
    process_args(args)
    myfile = str(Path(__file__).parent / 'testdir' / 'wordlist.tsv')
    testargs = ['', '--read-file', myfile, '-l', 'de', '--verbose']
    with patch.object(sys, 'argv', testargs):
        args = parse_args(testargs)
    process_args(args)


def test_filters():
    "Test functions for filtering of vocabulary."
    # init
    myvocab = load_wordlist(str(Path(__file__).parent / 'inputfile.txt'))
    assert len(myvocab) == 3

    # combined
    newvocab = combined_filters(deepcopy(myvocab), 'unknown')
    # assert newvocab == myvocab

    # input
    assert is_relevant_input('testing')
    assert is_relevant_input('testing.')
    assert not is_relevant_input('')
    assert not is_relevant_input('123')
    assert not is_relevant_input('abcde12345')
    assert not is_relevant_input('ABCDEF')

    # lemmata
    newvocab = recognized_by_simplemma(deepcopy(myvocab), DE_LEMMADATA)
    assert newvocab == {}

    # filters
    newvocab = sources_freqfilter(deepcopy(myvocab))
    assert len(newvocab) == 2
    newvocab = shortness_filter(deepcopy(newvocab))
    assert len(newvocab) == 1
    newvocab = sources_filter(deepcopy(myvocab), Counter(['Source1']))
    assert len(newvocab) == 1
    newvocab = wordlist_filter(deepcopy(myvocab), ['Tests', 'Other'], keep_words=False)
    assert len(newvocab) == 1
    newvocab = wordlist_filter(deepcopy(myvocab), ['Tests', 'Other'], keep_words=True)
    assert len(newvocab) == 2

    # scoring function
    scores = {}
    scores = scoring_func(scores, 1, newvocab)
    assert scores == {'Other': 1, 'Tests': 1}
    scores = scoring_func(scores, 1, newvocab)
    assert scores == {'Other': 2, 'Tests': 2}

    # hyphens
    myvocab = {
        'de-hyphening': ToEntry({'time_series': array('H', [1, 2, 3])}),
        'hyphens-stuff': ToEntry({'time_series': array('H', [3, 3])}),
        'hyphen-stuff': ToEntry({'time_series': array('H', [3, 3, 3])}),
        'stuff': ToEntry({'time_series': array('H', [3, 3, 3, 3, 3])})
    }
    myvocab = convert_to_numpy(myvocab)
    newvocab = hyphenated_filter(myvocab, perc=0, verbose=False)
    assert list(newvocab.keys()) == ['de-hyphening', 'stuff']

    # zipf
    myvocab = {
        'short': ToEntry({'time_series': array('H', [1, 2, 3])}),
        'the-longest-word': ToEntry({'time_series': array('H', [1, 2, 3])}),
        'longer': ToEntry({'time_series': array('H', [1])}),
        'other1': ToEntry({'time_series': array('H', [1, 2, 3])}),
        'this': ToEntry({'time_series': array('H', [1, 2])}),
        'one': ToEntry({'time_series': array('H', [1, 2, 3, 4, 5])}),
    }
    myvocab = convert_to_numpy(myvocab)
    newvocab = zipf_filter(deepcopy(myvocab), freqperc=80, lenperc=20, verbose=True)
    assert len(newvocab) == 5
    newvocab = zipf_filter(deepcopy(myvocab), freqperc=20, lenperc=20, verbose=True)
    assert len(newvocab) == 4

    # headings
    myvocab = {
        'Berg': ToEntry({'time_series': array('H', [1, 2, 3]), 'headings': True}),
        'Tal': ToEntry({'time_series': array('H', [1, 2, 3]), 'headings': False}),
    }
    myvocab = headings_filter(myvocab)
    assert len(myvocab) == 1

    # n-grams
    myvocab = dict.fromkeys(['Berg', 'Berge', 'Bergen', 'Berger', 'Bernstein', 'Tal', 'Täler'])
    newvocab = ngram_filter(deepcopy(myvocab), threshold=50, verbose=True)
    assert list(newvocab.keys()) == ['Berg', 'Berge', 'Berger', 'Bernstein', 'Tal']
    newvocab = ngram_filter(deepcopy(myvocab), threshold=10, verbose=True)
    assert list(newvocab.keys()) == ['Berge', 'Bernstein', 'Tal']



#def test_readme():
#    """Test function to verify readme examples."""
#    # ...
#


shoten/cli.py
"""Command-line interface."""

import argparse
import sys

from .shoten import gen_wordlist, load_wordlist
from .filters import combined_filters



def parse_args(args):
    'Parse and return CLI arguments.'
    parser = argparse.ArgumentParser(description='Command-line interface for Shoten')
    parser.add_argument("-f", "--read-file",
                        help="name of input file",
                        type=str)
    parser.add_argument("-d", "--read-dir",
                        help="name of input directory",
                        type=str)
    parser.add_argument("-l", "--language",
                        help="languages of interest",
                        nargs='+', default=[])
    parser.add_argument('--filter-level',
                        help="Choose a level of filtering",
                        choices=['loose', 'normal', 'strict'],
                        default='normal')
    parser.add_argument('-v', '--verbose', action="store_true",
                        help="toggle verbose output",
                        #action='count', default=0,
                        #help="increase logging verbosity (-v or -vv)",
                        )
    return parser.parse_args()


def process_args(args):
    'Process input according to CLI arguments.'
    if args.read_file:
        myvocab = load_wordlist(args.read_file, langcodes=args.language)
    elif args.read_dir:
        myvocab = gen_wordlist(args.read_dir, langcodes=args.language)
    try:
        myvocab = combined_filters(myvocab, args.filter_level)
    except IndexError:
        myvocab = {}
    for wordform in sorted(myvocab):
        if args.verbose is True:
            print(wordform, myvocab[wordform])
        else:
            print(wordform)


def main():
    """ Run as a command-line utility. """
    args = parse_args(sys.argv[1:])
    process_args(args)


if __name__ == '__main__':
    main()


shoten/datatypes.py
"""Module listing data classes, types and constants used by shoten."""


#from __future__ import annotations

from array import array
from collections import Counter
from datetime import datetime


TODAY = datetime.today()


ARRAY_TYPE = 'H'
MAX_SERIES_VAL = 65535

MAX_NGRAM_VOC = 15000


class Entry:
    "Defines a class for dictionaries entries, containing metadata and stats."
    __slots__ = ['absfreq', 'headings', 'mean', 'series_abs', 'series_rel', 'sources', 'stddev', 'time_series', 'total']
    def __init__(self):
        self.absfreq: int
        self.headings: bool = False
        self.mean: float
        self.series_abs = array('f')
        self.series_rel = array('f')
        self.sources = Counter()
        self.stddev: float
        self.time_series = array(ARRAY_TYPE)
        self.total: int


shoten/filters.py
"""Filters meant to reduce noise."""


import csv
import re
import string

from collections import Counter
#from copy import deepcopy
from math import ceil

import numpy as np

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

from simplemma import is_known, lemmatize

from .datatypes import MAX_NGRAM_VOC, MAX_SERIES_VAL




RE_FILTER = re.compile(r'[^\W\d\.-]')



def print_changes(phase, old_len, new_len):
    'Report on absolute and relative changes.'
    coeff = 100 - (new_len/old_len)*100
    print(phase, 'removed', old_len - new_len, '%.1f' % coeff, '%')


def scoring_func(scores, value, newvocab):
    'Defines scores for each word and add values.'
    for wordform in set(newvocab):
        if wordform not in scores:
            scores[wordform] = 0
        scores[wordform] += value
    return scores


def store_results(myvocab, filename):
    'Write vocabulary with essential data to file.'
    with open(filename, 'w', encoding='utf-8') as outfile:
        tsvwriter = csv.writer(outfile, delimiter='\t')
        tsvwriter.writerow(['word', 'sources', 'time series'])
        # for token in sorted(myvocab, key=locale.strxfrm):
        for entry in myvocab:
            tsvwriter.writerow([entry, ','.join(myvocab[entry].sources), str(myvocab[entry].time_series.tolist())])


def combined_filters(myvocab, setting):
    '''Apply a combination of filters based on the chosen setting.'''
    if setting == 'loose':
        return freshness_filter(frequency_filter(hapax_filter(myvocab)))
    if setting == 'normal':
        return freshness_filter(oldest_filter(frequency_filter(hapax_filter(myvocab))))
    if setting == 'strict':
        return freshness_filter(oldest_filter(
               frequency_filter(shortness_filter(ngram_filter(hapax_filter(myvocab))), min_perc=10)
               ))
    return myvocab


def is_relevant_input(token):
    'Determine if the given token is to be considered relevant for further processing.'
    # apply filter first
    if not 5 <= len(token) <= 50 or token.startswith('@') or token.startswith('#') or token.endswith('-'):
        return False
    token = token.rstrip(string.punctuation)
    if len(token) == 0 or token.isnumeric() or not RE_FILTER.search(token):
        return False
    if sum(map(str.isupper, token)) > 4 or sum(map(str.isdigit, token)) > 4:
        return False
    return True


def hapax_filter(myvocab, freqcount=2):
    '''Eliminate hapax legomena and delete same date only.'''
    old_len = len(myvocab)
    for token in [t for t in myvocab if np.unique(myvocab[t].time_series).shape[0] <= freqcount]:
        del myvocab[token]
    print_changes('sameness/hapax', old_len, len(myvocab))
    return myvocab


def recognized_by_simplemma(myvocab, lemmadata):
    'Run the simplemma lemmatizer to check if input is recognized.'
    old_len = len(myvocab)
    for token in [t for t in myvocab if is_known(t, lemmadata)]:
        del myvocab[token]
    print_changes('known by simplemma', old_len, len(myvocab))
    deletions = []
    for word in myvocab:
        try:
            lemmatize(word, lemmadata, greedy=True, silent=False)
        except ValueError:
            deletions.append(word)
    for token in deletions:
        del myvocab[token]
    print_changes('reduced by simplemma', old_len, len(myvocab))
    return myvocab


def shortness_filter(myvocab, threshold=20):
    '''Eliminate short words'''
    old_len = len(myvocab)
    lengths = np.array([len(l) for l in myvocab])
    lenthreshold = np.percentile(lengths, threshold)
    for token in [t for t in myvocab if len(t) < lenthreshold]:
        del myvocab[token]
    print_changes('short words', old_len, len(myvocab))
    return myvocab


def frequency_filter(myvocab, max_perc=50, min_perc=.001): # 50 / 0.01
    '''Reduce dict size by stripping least and most frequent items.'''
    old_len = len(myvocab)
    myfreq = np.array([myvocab[l].time_series.shape[0] for l in myvocab])
    min_thres, max_thres = np.percentile(myfreq, min_perc), np.percentile(myfreq, max_perc)
    for token in [t for t in myvocab if
                  myvocab[t].time_series.shape[0] < min_thres or myvocab[t].time_series.shape[0] > max_thres
                 ]:
        del myvocab[token]
    print_changes('most/less frequent', old_len, len(myvocab))
    return myvocab


def hyphenated_filter(myvocab, perc=50, verbose=False): # threshold in percent
    '''Reduce dict size by deleting hyphenated tokens when the parts are frequent.'''
    deletions, old_len = [], len(myvocab)
    myfreqs = np.array([myvocab[l].time_series.shape[0] for l in myvocab])
    threshold = np.percentile(myfreqs, perc)
    for word in [w for w in myvocab if '-' in w]:
        mylist = word.split('-')
        firstpart, secondpart = mylist[0], mylist[1]
        if firstpart in myvocab and myvocab[firstpart].time_series.shape[0] > threshold or \
           secondpart in myvocab and myvocab[secondpart].time_series.shape[0] > threshold:
            deletions.append(word)
    if verbose is True:
        print(sorted(deletions))
    for item in deletions:
        del myvocab[item]
    print_changes('hyphenated', old_len, len(myvocab))
    return myvocab


def oldest_filter(myvocab, threshold=50):
    '''Reduce number of candidate by stripping the oldest.'''
    # todo: what about cases like [365, 1, 1, 1] ?
    old_len = len(myvocab)
    myratios = np.array([np.sum(myvocab[l].time_series)/myvocab[l].time_series.shape[0] for l in myvocab])
    threshold = np.percentile(myratios, threshold)
    for token in [t for t in myvocab if np.sum(myvocab[t].time_series)/myvocab[t].time_series.shape[0] > threshold]:
        del myvocab[token]
    print_changes('oldest', old_len, len(myvocab))
    return myvocab


def zipf_filter(myvocab, freqperc=65, lenperc=35, verbose=False):
    '''Filter candidates based on a approximation of Zipf's law.'''
    # todo: count length without hyphen or punctuation: len(l) - l.count('-')
    old_len = len(myvocab)
    freqs = np.array([myvocab[l].time_series.shape[0] for l in myvocab])
    freqthreshold = np.percentile(freqs, freqperc)
    lengths = np.array([len(l) for l in myvocab])
    lenthreshold = np.percentile(lengths, lenperc)
    deletions = []
    for token in [t for t in myvocab if myvocab[t].time_series.shape[0] >= freqthreshold and len(t) <= lenthreshold]:
        deletions.append(token)
        # if verbose is True:
            # print(token, len(token), myvocab[token].time_series.shape[0])
        del myvocab[token]
    if verbose is True:
        print(sorted(deletions))
    print_changes('zipf frequency', old_len, len(myvocab))
    return myvocab


def freshness_filter(myvocab, percentage=10):
    '''Define a freshness threshold to model series of token occurrences in time.'''
    old_len = len(myvocab)
    mysums = np.array([np.sum(myvocab[l].time_series) for l in myvocab])
    datethreshold = np.percentile(mysums, percentage)
    deletions = []
    for token in myvocab:
        # re-order
        myvocab[token].time_series = -np.sort(-myvocab[token].time_series)
        # thresholds
        thresh = myvocab[token].time_series.shape[0]*(percentage/100)
        freshnessindex = np.sum(myvocab[token].time_series[-ceil(thresh):])
        oldnessindex = np.sum(myvocab[token].time_series[:ceil(thresh)])
        if oldnessindex < datethreshold:
            #if oldnessindex < np.percentile(myvocab[token].time_series, percentage):
            #    continue
            thresh = myvocab[token].time_series.shape[0]*(percentage/100)
            freshnessindex = np.sum(myvocab[token].time_series[-ceil(thresh):])
            if freshnessindex < np.percentile(myvocab[token].time_series, percentage):
                deletions.append(token)
            # print(myvocab[token], freshnessindex, oldnessindex, token)
    for item in deletions:
        del myvocab[item]
    print_changes('freshness', old_len, len(myvocab))
    return myvocab


def sources_freqfilter(myvocab, threshold=2, balanced=True):
    '''Filter words based on source diversity.'''
    deletions = []
    i, j = 0, 0
    for word in myvocab:
        if len(myvocab[word].sources) == 0:
            continue
        # absolute number
        if len(myvocab[word].sources) < threshold:
            deletions.append(word)
            i += 1
            continue
        # distribution of sources
        if balanced is True:
            values = [t[1] for t in myvocab[word].sources.most_common()]
            # first value too present compared to the rest
            if values[0] >= 4*values[1]: # (sum(values)/len(values)):
                deletions.append(word)
                j += 1
                continue
    old_len = len(myvocab)
    for item in deletions:
        del myvocab[item]
    print_changes('sources freq', old_len, old_len-i)
    print_changes('sources balance', old_len, old_len-j)
    return myvocab


def sources_filter(myvocab, myset):
    '''Only keep the words for which the source contains at least
       one string listed in the input set.'''
    deletions = []
    for word in myvocab:
        deletion_flag = True
        if len(myvocab[word].sources) > 0:
            for source in myvocab[word].sources:
                # for / else construct
                for mystring in myset:
                    if mystring in source:
                        deletion_flag = False
                        break
                else:
                    continue
                # inner loop was broken, break the outer
                break
        # record deletion
        if deletion_flag:
            deletions.append(word)
    old_len = len(myvocab)
    for item in deletions:
        del myvocab[item]
    print_changes('sources list', old_len, len(myvocab))
    return myvocab


def wordlist_filter(myvocab, mylist, keep_words=False):
    '''Keep or discard words present in the input list.'''
    mylist = set(mylist)
    intersection = [w for w in myvocab if w in mylist]
    if keep_words is False:
        deletions = intersection
    else:
        intersection = set(intersection)
        deletions = [w for w in myvocab if w not in intersection]
    old_len = len(myvocab)
    for word in deletions:
        del myvocab[word]
    print_changes('word list', old_len, len(myvocab))
    return myvocab


def headings_filter(myvocab):
    '''Filter words based on their presence in headings.'''
    deletions = [word for word in myvocab if myvocab[word].headings is False]
    old_len = len(myvocab)
    for item in deletions:
        del myvocab[item]
    print_changes('headings', old_len, len(myvocab))
    return myvocab


def read_freqlist(filename):
    'Read frequency info from a TSV file.'
    freqlimits = {}
    with open(filename, 'r', encoding='utf-8') as csvfile:
        tsvreader = csv.reader(csvfile, delimiter='\t')
        # skip headline
        next(tsvreader)
        for row in tsvreader:
            # unpack
            word, mean, stddev = row[0], float(row[2]), float(row[3])
            # limit
            freqlimits[word] = float('{0:.3f}'.format(mean + 2*stddev))
    return freqlimits


def longtermfilter(myvocab, filename, mustexist=False, startday=1, interval=7):
    'Discard words which are not significantly above a mean long-term frequency.'
    freqlimits = read_freqlist(filename)
    oldestday = startday + interval - 1
    allfreqs = 0
    for word in myvocab:
        mydays = Counter(myvocab[word].time_series)
        occurrences = sum(
            mydays[day]
            for day in range(oldestday, startday - 1, -1)
            if day in mydays
        )
        # compare with maximum possible value
        occurrences = min(MAX_SERIES_VAL, occurrences)
        # compute totals
        myvocab[word].absfreq = occurrences
        allfreqs += occurrences
    # relative frequency
    deletions = []
    intersection = [w for w in myvocab if w in freqlimits]
    for word in intersection:
        relfreq = (myvocab[word].absfreq / allfreqs)*1000000
        # threshold defined by long-term frequencies
        if relfreq < freqlimits[word]:
            #print(word, relfreq, freqlimits[word])
            deletions.append(word)
    if mustexist is True:
        deletions += [w for w in myvocab if w not in freqlimits]
    old_len = len(myvocab)
    for item in deletions:
        del myvocab[item]
    print_changes('long-term frequency threshold', old_len, len(myvocab))
    return myvocab


def ngram_filter(myvocab, threshold=90, verbose=False):
    '''Find dissimilar tokens based on character n-gram occurrences.'''
    lengths = np.array([len(l) for l in myvocab])
    minlengthreshold = np.percentile(lengths, 1)
    for i in (70, 65, 60, 55, 50, 45, 40):
        maxlengthreshold = np.percentile(lengths, i)
        mytokens = [t for t in myvocab if minlengthreshold <= len(t) <= maxlengthreshold]
        #print(i, len(mytokens))
        if len(mytokens) <= MAX_NGRAM_VOC:
            break
    if len(mytokens) > MAX_NGRAM_VOC:
        print('Vocabulary size too large, skipping n-gram filtering')
        return myvocab
    old_len = len(myvocab)
    # token cosine similarity
    max_exp = 21
    vectorizer = CountVectorizer(analyzer='char', max_features=2 ** max_exp, ngram_range=(1,4), strip_accents=None, lowercase=True, max_df=1.0)
    firstset = set(compute_deletions(mytokens, vectorizer, threshold))
    vectorizer = TfidfVectorizer(analyzer='char', max_features=2 ** max_exp, ngram_range=(1,4), strip_accents=None, lowercase=True, max_df=1.0, sublinear_tf=True, binary=True)
    secondset = set(compute_deletions(mytokens, vectorizer, threshold))
    for token in firstset.intersection(secondset):
        del myvocab[token]
    if verbose is True:
        print(sorted(firstset.intersection(secondset)))
    print_changes('ngrams', old_len, len(myvocab))
    return myvocab


def compute_deletions(mytokens, vectorizer, threshold):
    '''Compute deletion list based on n-gram dissimilarities.'''
    count_matrix = vectorizer.fit_transform(mytokens)
    cosine_similarities = cosine_similarity(count_matrix)  # linear_kernel, laplacian_kernel, rbf_kernel, sigmoid_kernel
    myscores = {
        # np.median(cosine_similarities[:, rownum]) * (np.log(len(mytokens[rownum]))/len(mytokens[rownum]))
        mytokens[rownum]: 1 - np.mean(cosine_similarities[:, rownum])
        for rownum, _ in enumerate(mytokens)
    }

    # print info
    #if verbose is True:
    #    resultsize = 20 # :resultsize*2
    #    for k, v in sorted(myscores.items(), key=lambda item: item[1], reverse=True)[:resultsize]:
    #        print(k, v)
    #    #print()
    #    for k, v in sorted(myscores.items(), key=lambda item: item[1])[:resultsize]:
    #        print(k, v)
    # process
    mylist = np.array([myscores[s] for s in myscores])
    return [s for s in myscores if myscores[s] >= np.percentile(mylist, threshold)]
